File: DebateUser/DebateUserBackEnd/src/main/java/com/debate/service/LikeService.java
package com.debate.service;

import com.debate.entity.Debate;
import com.debate.entity.Like;
import com.debate.entity.User;
import com.debate.exception.ResourceNotFoundException;
import com.debate.repository.DebateRepository;
import com.debate.repository.LikeRepository;
import com.debate.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class LikeService {
    private final LikeRepository likeRepository;
    private final DebateRepository debateRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;

    @Transactional
    public void toggleLike(Long debateId, Long userId) {
        likeRepository.findByDebateIdAndUserId(debateId, userId)
                .ifPresentOrElse(
                        likeRepository::delete,
                        () -> {
                            Debate debate = debateRepository.findById(debateId)
                                    .orElseThrow(() -> new ResourceNotFoundException("토론을 찾을 수 없습니다"));
                            User user = userRepository.getReferenceById(userId);
                            
                            Like like = Like.builder()
                                    .debate(debate)
                                    .user(user)
                                    .build();
                            likeRepository.save(like);

                            // 알림 생성 로직 (본인이 아닐 경우)
                            if (!debate.getUser().getId().equals(userId)) {
                                try {
                                    notificationService.createNotification(
                                            debate.getUser(),
                                            user.getNickname() + "님이 회원님의 토론을 좋아합니다: " + debate.getTitle(),
                                            "LIKE",
                                            "/debate/" + debate.getId()
                                    );
                                } catch (Exception e) {
                                    System.err.println("알림 생성 실패: " + e.getMessage());
                                }
                            }
                        }
                );
    }

    public boolean isLiked(Long debateId, Long userId) {
        System.out.println("LikeService.isLiked - DebateID: " + debateId + ", UserID: " + userId);
        boolean exists = likeRepository.existsByDebateIdAndUserId(debateId, userId);
        System.out.println("LikeService.isLiked - Exists: " + exists);
        return exists;
    }
}

File: DebateUser/DebateUserFrontEnd/src/services/likeService.js
/**
 * 좋아요 서비스 (Like Service)
 * 
 * 토론 좋아요 관련 API 호출을 담당하는 서비스입니다.
 * 
 * 주요 기능:
 * - 토론 좋아요 토글 (좋아요 추가/제거)
 * - 토론 좋아요 여부 확인
 */

import api from './api'

export const likeService = {
  /**
   * 좋아요 토글
   * 
   * 토론에 좋아요를 추가하거나 제거합니다. 인증이 필요합니다.
   * 이미 좋아요를 누른 경우 제거되고, 누르지 않은 경우 추가됩니다.
   * 
   * @param {number} debateId - 토론 ID
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response.data - 좋아요 상태 정보
   */
  async toggleLike(debateId) {
    const response = await api.post(`/likes/debate/${debateId}`)
    return response.data
  },

  /**
   * 좋아요 여부 확인
   * 
   * 현재 로그인한 사용자가 특정 토론에 좋아요를 눌렀는지 확인합니다.
   * 인증이 필요합니다.
   * 
   * @param {number} debateId - 토론 ID
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {boolean} response.data - 좋아요 여부 (true/false)
   */
  async isLiked(debateId) {
    const response = await api.get(`/likes/debate/${debateId}`)
    return response.data
  },
}

